File: sim_harness/core/test_isolation.py
# Copyright 2026 The sim_harness Authors
# SPDX-License-Identifier: Apache-2.0

"""
Test isolation utilities for ROS 2 tests.

Prevents crosstalk between concurrent tests by assigning unique
ROS_DOMAIN_ID values and Gazebo partitions.
"""

import os
import random
from dataclasses import dataclass
from typing import Optional


@dataclass
class TestIsolationConfig:
    """Configuration for test isolation."""

    domain_id: int
    """ROS 2 domain ID (0-232)."""

    gz_partition: str
    """Gazebo partition name for simulation isolation."""


def get_test_isolation_config() -> TestIsolationConfig:
    """
    Get test isolation configuration.

    Reads ROS_DOMAIN_ID from environment and maps it to a test-safe
    range (100-199) to prevent collision with production systems.

    Returns:
        TestIsolationConfig with domain ID and Gazebo partition
    """
    # Read domain ID from environment, default to 0
    env_domain = os.environ.get('ROS_DOMAIN_ID', '0')
    try:
        domain_id = int(env_domain)
    except ValueError:
        domain_id = 0

    # Map to test isolation range (100-199)
    test_domain_id = 100 + (domain_id % 100)

    # Generate Gazebo partition name
    gz_partition = f"gz_test_{test_domain_id}"

    return TestIsolationConfig(
        domain_id=test_domain_id,
        gz_partition=gz_partition
    )


def apply_test_isolation(config: Optional[TestIsolationConfig] = None) -> TestIsolationConfig:
    """
    Apply test isolation by setting environment variables.

    Sets ROS_DOMAIN_ID and GZ_PARTITION environment variables to
    isolate the test from other ROS 2 systems.

    Args:
        config: Optional config to use (generates new if None)

    Returns:
        The applied TestIsolationConfig
    """
    if config is None:
        config = get_test_isolation_config()

    os.environ['ROS_DOMAIN_ID'] = str(config.domain_id)
    os.environ['GZ_PARTITION'] = config.gz_partition

    return config


def generate_test_node_name(
    base_name: str = "test_node",
    domain_id: Optional[int] = None
) -> str:
    """
    Generate a unique node name for test isolation.

    Combines base name with domain suffix and random number to
    prevent node name collisions.

    Args:
        base_name: Base node name
        domain_id: Optional domain ID (reads from env if None)

    Returns:
        Unique node name like "test_node_d100_1234"
    """
    if domain_id is None:
        config = get_test_isolation_config()
        domain_id = config.domain_id

    random_suffix = random.randint(1000, 9999)
    return f"{base_name}_d{domain_id}_{random_suffix}"

File: src/core/test_isolation.cpp
// Copyright 2026 John Vial
// SPDX-License-Identifier: Apache-2.0

#include "sim_harness/core/test_isolation.hpp"

#include <cstdlib>
#include <random>
#include <sstream>

namespace sim_harness
{

TestIsolationConfig getTestIsolationConfig()
{
  TestIsolationConfig config;

  // Get ROS_DOMAIN_ID from environment (default to 0)
  const char * domain_env = std::getenv("ROS_DOMAIN_ID");
  int domain_id = domain_env ? std::stoi(domain_env) : 0;

  // Map to valid range (100-199 for test isolation)
  int mapped_domain;
  if (domain_id >= 100 && domain_id <= 199) {
    // Already in valid range
    mapped_domain = domain_id;
  } else if (domain_id >= 200) {
    // Map 200+ to 100-199
    mapped_domain = 100 + ((domain_id - 200) % 100);
  } else {
    // Map 0-99 to 100-199
    mapped_domain = 100 + (domain_id % 100);
  }

  config.domain_id = std::to_string(mapped_domain);
  config.simulator_partition = "gz_test_" + config.domain_id;

  return config;
}

void applyTestIsolation(const TestIsolationConfig & config)
{
  // Set ROS domain ID
  setenv("ROS_DOMAIN_ID", config.domain_id.c_str(), 1);

  // Set Gazebo partition (for process isolation)
  setenv("GZ_PARTITION", config.simulator_partition.c_str(), 1);
}

std::string generateTestNodeName(const std::string & base_name)
{
  // Get domain ID for uniqueness
  const char * domain_env = std::getenv("ROS_DOMAIN_ID");
  std::string domain_suffix = domain_env ? domain_env : "0";

  // Add random suffix for additional uniqueness
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_int_distribution<> dis(1000, 9999);

  std::stringstream ss;
  ss << base_name << "_d" << domain_suffix << "_" << dis(gen);

  return ss.str();
}

}  // namespace sim_harness
